atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+                return ImmutableSequenceAndBound.of(seq, PtBytes.toLong(value));
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
+                long backoffTime = getRandomBackoffTime();
+                try {
+                    Thread.sleep(backoffTime);
+                } catch (InterruptedException e1) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private Optional<SequenceAndBound> getLearnedState(long seq) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return Optional.of(ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L));
+        }
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(learners),
+                l -> getLearnedValue(seq, l),
+                1,
+                executor,
+                PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS,
+                true);
+        if (responses.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(ImmutableSequenceAndBound.of(seq, responses.iterator().next().getValue()));
+    }
+
+    private static PaxosLong getLearnedValue(long seq, PaxosLearner learner) {
+        PaxosValue value = learner.getLearnedValue(seq);
+        if (value == null) {
+            throw new NoSuchElementException();
+        }
+        return ImmutablePaxosLong.of(PtBytes.toLong(value.getData()));
+    }
+
+    @Override
+    public synchronized void storeUpperLimit(long limit) throws MultipleRunningTimestampServiceError {
+        Preconditions.checkArgument(agreedState == null || limit >= agreedState.getBound());
+        long newSeq = agreedState == null ? 0 : agreedState.getSeqId() + 1;
+        while (true) {
+            try {
+                proposer.propose(newSeq, PtBytes.toBytes(limit));
+                PaxosValue value = knowledge.getLearnedValue(newSeq);
+                if (!value.getLeaderUUID().equals(proposer.getUUID())) {
+                    throw new MultipleRunningTimestampServiceError("concurrent proposal: " + value);
+                }
+                long newLimit = PtBytes.toLong(value.getData());
+                agreedState = ImmutableSequenceAndBound.of(newSeq, newLimit);
+                if (newLimit >= limit) {
+                    return;
+                }
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
+                long backoffTime = getRandomBackoffTime();
+                try {
+                    wait(backoffTime);
+                } catch (InterruptedException e1) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private long getRandomBackoffTime() {
+        return (long) (randomWaitBeforeProposalMs * Math.random());
```

Comment:
minimum = 1 is fine for now, I think. Since `Math.random()` can't actually return 1, as you say the maximum is still respected, so I've just added 1.